File: android-app/src/main/java/fr/cph/chicago/data/MarkerDataHolder.java
package fr.cph.chicago.data;

import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;

import java.util.HashMap;
import java.util.Map;

import fr.cph.chicago.entity.AStation;

public class MarkerDataHolder {
    private final Map<LatLng, MarkerDataHolder.MarkerHolder> data;

    public MarkerDataHolder() {
        data = new HashMap<>();
    }

    public void addData(final Marker marker, final AStation station) {
        final MarkerDataHolder.MarkerHolder markerHolder = new MarkerDataHolder.MarkerHolder();
        markerHolder.setMarker(marker);
        markerHolder.setStation(station);
        final LatLng latLng = marker.getPosition();
        data.put(latLng, markerHolder);
    }

    public void clear() {
        data.clear();
    }

    public AStation getStation(final Marker marker) {
        return data.get(marker.getPosition()).getStation();
    }

    private class MarkerHolder {
        private Marker marker;
        private AStation station;

        public AStation getStation() {
            return station;
        }

        public void setStation(AStation station) {
            this.station = station;
        }

        public Marker getMarker() {
            return marker;
        }

        public void setMarker(Marker marker) {
            this.marker = marker;
        }
    }
}

File: android-app/src/main/java/fr/cph/chicago/data/MarkerDataHolder.kt
package fr.cph.chicago.data

import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import fr.cph.chicago.entity.AStation
import java.util.*

class MarkerDataHolder {
    private val data: MutableMap<LatLng, MarkerDataHolder.MarkerHolder>

    init {
        data = HashMap()
    }

    fun addData(marker: Marker, station: AStation) {
        val markerHolder = MarkerDataHolder.MarkerHolder()
        markerHolder.marker = marker
        markerHolder.station = station
        val latLng = marker.position
        data.put(latLng, markerHolder)
    }

    fun clear() {
        data.clear()
    }

    fun getStation(marker: Marker): AStation? {
        return data[marker.position]!!.station
    }

    private class MarkerHolder {
        var marker: Marker? = null
        var station: AStation? = null
    }
}
